### PR TITLE
Config check ssd

### PR DESCRIPTION
**Pull request by AndreMouche**
hi, this PR check data device's rotational info before the start. @BusyJay @zhangjinpeng1987 PTAL

### CODE REVIEW

File: src/util/config.rs [hunk 1]
```diff
@@ -739,6 +739,132 @@ pub fn check_kernel() -> Vec<ConfigError> {
     Vec::new()
 }
 
+#[derive(Debug, Default)]
```
Comment (siddontang) [inline src/util/config.rs]: can we move this to another file, it is crowded in the config file now. 

Maybe we should create a config directory now.

Comment (AndreMouche) [reply src/util/config.rs]: Maybe we could refactor the config file into a directory in another PR?

File: src/bin/tikv-server.rs [hunk 2]
```diff
@@ -145,6 +145,10 @@ fn check_system_config(config: &TiKvConfig) {
         env::set_var("TZ", ":/etc/localtime");
         warn!("environment variable `TZ` is missing, use `/etc/localtime`");
     }
+
+    if let Err(e) = util::config::check_data_dir(&config.storage.data_dir) {
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: I remember that the path of raft db and kv db can be configured.

File: src/util/config.rs [hunk 3]
```diff
@@ -739,6 +739,132 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+struct FsInfo {
+    tp: String,
+    opts: String,
+    mnt_dir: String,
+    fsname: String,
+}
+
+#[cfg(target_os = "linux")]
+fn get_fsinfo(path: &str, mnt_file: &str) -> Result<FsInfo, ConfigError> {
```
Comment (BusyJay) [inline src/util/config.rs]: s/get_fsinfo/get_fs_info

File: src/util/config.rs [hunk 4]
```diff
@@ -739,6 +739,132 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+        let profile = CString::new(mnt_file).unwrap();
+        let retype = CString::new("r").unwrap();
+        let afile = libc::setmntent(profile.as_ptr(), retype.as_ptr());
+        let mut fs = FsInfo::default();
+        loop {
+            let ent = libc::getmntent(afile);
+            if ent.is_null() {
+                break;
+            }
+            let cur_dir = CStr::from_ptr((*ent).mnt_dir).to_str().unwrap();
```
Comment (BusyJay) [inline src/util/config.rs]: Can you reassign it as `let ent = &* ent`?

File: src/util/config.rs [hunk 5]
```diff
@@ -739,6 +739,132 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+            )));
+        }
+        Ok(fs)
+    }
+}
+
+pub fn get_rotational_info(fsname: &str) -> Result<String, ConfigError> {
+    if !fsname.starts_with("/dev/") {
+        return Err(ConfigError::Limit(format!(
+            "fsname:{:?} is not a normal device",
```
Comment (BusyJay) [inline src/util/config.rs]: You mean block device?

File: src/util/config.rs [hunk 6]
```diff
@@ -739,6 +739,132 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+            return Err(ConfigError::Limit(format!(
+                "fs:{} no device find in block",
+                fsname
+            )));
+        }
+        let cur_path = format!("/sys/block/{}", dev);
+        if Path::new(&cur_path).exists() {
+            break;
+        }
+        dev = &dev[0..dev.len() - 1];
```
Comment (BusyJay) [inline src/util/config.rs]: Why?

File: src/util/config.rs [hunk 7]
```diff
@@ -739,6 +739,132 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+        )));
+    }
+    use std::fs::File;
+    use std::io::Read;
+
+    let mut buffer = String::new();
+    File::open(rpath)
+        .and_then(|mut f| f.read_to_string(&mut buffer))
+        .map_err(|e| {
+            ConfigError::Limit(format!(
```
Comment (BusyJay) [inline src/util/config.rs]: Why is it a `Limit` error?

File: src/util/config.rs [hunk 8]
```diff
@@ -739,6 +739,132 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+            return Err(ConfigError::Limit(format!(
+                "path:{:?} not find in mountable",
+                path
+            )));
+        }
+        Ok(fs)
+    }
+}
+
+pub fn get_rotational_info(fsname: &str) -> Result<String, ConfigError> {
```
Comment (BusyJay) [inline src/util/config.rs]: Why `pub`?

File: src/util/config.rs (hunk 6 above)
Comment (AndreMouche) [reply src/util/config.rs]: Since `/dev/sda2` and `/dev/sda1` 's block info  maybe in directory `/sys/block/sda/`.

File: src/util/config.rs [hunk 9]
```diff
@@ -739,6 +743,142 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+        }
+        Ok(fs)
+    }
+}
+
+#[cfg(target_os = "linux")]
+fn get_rotational_info(fsname: &str) -> Result<String, ConfigError> {
+    if !fsname.starts_with("/dev/") {
+        return Err(ConfigError::FileSystem(format!(
+            "fsname:{:?} is not a normal block device",
```
Comment (BusyJay) [inline src/util/config.rs]: `normal` is misleading.

File: src/util/config.rs [hunk 10]
```diff
@@ -739,6 +743,142 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+    use std::path::Path;
+    let block_dir = "/sys/block";
+    let mut device_dir = format!("{}/{}", block_dir, dev);
+    if !Path::new(&device_dir).exists() {
+        let dir = fs::read_dir(&block_dir).map_err(|e| {
+            ConfigError::FileSystem(format!("read block dir {} with error:{:?}", block_dir, e))
+        })?;
+        let mut find = false;
+        for entry in dir {
+            if let Ok(entry) = entry {
```
Comment (BusyJay) [inline src/util/config.rs]: You can use `is_err` to reduce indent.

File: src/util/config.rs [hunk 11]
```diff
@@ -739,6 +743,142 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+            )))
+        }
+    };
+
+    let fs_info = get_fs_info(&real_path, "/proc/mounts")?;
+    // TODO check ext4 nodelalloc
+    info!("data_path:{}, mount fs info:{:?}", data_path, fs_info);
+    let rotational_info = get_rotational_info(&fs_info.fsname)?;
+    let msg = format!(
+        "data_path:{},device rational info {:?}",
```
Comment (BusyJay) [inline src/util/config.rs]: space.

File: src/util/config.rs [hunk 12]
```diff
@@ -739,6 +743,142 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+        data_path, rotational_info
+    );
+    if rotational_info == "0" {
+        // ssd
+        info!("{}", msg);
+    } else {
+        warn!("{}", msg);
```
Comment (BusyJay) [inline src/util/config.rs]: Confusing log.

File: src/util/config.rs [hunk 13]
```diff
@@ -739,6 +743,142 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+#[derive(Debug, Default)]
+struct FsInfo {
+    tp: String,
+    opts: String,
+    mnt_dir: String,
+    fsname: String,
+}
+
+#[cfg(target_os = "linux")]
+fn get_fs_info(path: &str, mnt_file: &str) -> Result<FsInfo, ConfigError> {
```
Comment (BusyJay) [inline src/util/config.rs]: Put things into module. Otherwise it becomes unused function on non-linux platform.

File: src/util/config.rs [hunk 14]
```diff
@@ -743,68 +743,76 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
-            if path.starts_with(&cur_dir) && cur_dir.len() >= fs.mnt_dir.len() {
-                fs.tp = CStr::from_ptr(ent.mnt_type).to_str().unwrap().to_owned();
-                fs.opts = CStr::from_ptr(ent.mnt_opts).to_str().unwrap().to_owned();
-                fs.fsname = CStr::from_ptr(ent.mnt_fsname).to_str().unwrap().to_owned();
-                fs.mnt_dir = cur_dir.to_owned();
+
+            libc::endmntent(afile);
+            if fs.mnt_dir.is_empty() {
+                return Err(ConfigError::FileSystem(format!(
+                    "path:{:?} not find in mountable",
```
Comment (BusyJay) [inline src/util/config.rs]: space.

File: src/util/config.rs [hunk 15]
```diff
@@ -743,68 +743,76 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+    }
 
-        libc::endmntent(afile);
-        if fs.mnt_dir.is_empty() {
+    pub fn get_rotational_info(fsname: &str) -> Result<String, ConfigError> {
+        if !fsname.starts_with("/dev/") {
             return Err(ConfigError::FileSystem(format!(
-                "path:{:?} not find in mountable",
-                path
+                "fsname:{:?} is not a device",
```
Comment (BusyJay) [inline src/util/config.rs]: space.

File: src/util/config.rs [hunk 16]
```diff
@@ -743,68 +743,76 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
-        let mut find = false;
-        for entry in dir {
-            if let Ok(entry) = entry {
+        let dev = fsname.trim_left_matches("/dev/");
+        use std::path::Path;
+        let block_dir = "/sys/block";
+        let mut device_dir = format!("{}/{}", block_dir, dev);
+        if !Path::new(&device_dir).exists() {
+            let dir = fs::read_dir(&block_dir).map_err(|e| {
+                ConfigError::FileSystem(format!("read block dir {} with error:{:?}", block_dir, e))
```
Comment (BusyJay) [inline src/util/config.rs]: Please leave a space after `:`.

File: src/util/config.rs [hunk 17]
```diff
@@ -813,67 +821,67 @@ fn get_rotational_info(fsname: &str) -> Result<String, ConfigError> {
...
-        Err(e) => {
-            return Err(ConfigError::FileSystem(format!(
-                "path:{:?} canonicalize failed with error: {:?}",
-                data_path, e
-            )))
+        let fs_info = get_fs_info(&real_path, "/proc/mounts")?;
+        // TODO check ext4 nodelalloc
+        info!("data_path: {}, mount fs info:{:?}", data_path, fs_info);
+        let rotational_info = get_rotational_info(&fs_info.fsname)?;
+        let msg = format!(
```
Comment (BusyJay) [inline src/util/config.rs]: This is unnecessary, just `$path is not on SSD device` is OK.

File: src/util/config.rs [hunk 18]
```diff
@@ -1101,4 +1250,57 @@ mod test {
             assert_eq!(check_addr(addr).is_ok(), is_ok);
         }
     }
+
+    #[cfg(target_os = "linux")]
+    #[test]
+    fn test_get_fs_info() {
```
Comment (BusyJay) [inline src/util/config.rs]: You can put the cases into the module above too.

**Comment by AndreMouche** [comment]
PTAL

**Comment by AndreMouche** [comment]
@BusyJay @zhangjinpeng1987 PTAL

File: src/util/config.rs [hunk 19]
```diff
@@ -739,6 +743,199 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+
+    fn get_rotational_info(fsname: &str) -> Result<String, ConfigError> {
+        if !fsname.starts_with("/dev/") {
+            return Err(ConfigError::FileSystem(format!(
+                "fsname: {:?} is not a block device",
+                fsname
+            )));
+        }
+        let dev = fsname.trim_left_matches("/dev/");
+        use std::path::Path;
```
Comment (BusyJay) [inline src/util/config.rs]: Move to module root.

File: src/util/config.rs [hunk 20]
```diff
@@ -739,6 +743,233 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+            // test normal case
+            let ret = check_data_dir("./", "/proc/mounts");
+            assert!(ret.is_ok());
+        }
+
+        #[test]
+        fn test_check_data_dir_with_device_mapper() {
+            let tmp_dir = TempDir::new("test-get-fs-info").unwrap();
+            let data_path = format!("{}/data1", tmp_dir.path().display());
+            canonicalize_path(&data_path).unwrap();
```
Comment (BusyJay) [inline src/util/config.rs]: This has no actual effect.

File: src/util/config.rs [hunk 21]
```diff
@@ -739,6 +743,233 @@ pub fn check_kernel() -> Vec<ConfigError> {
...
+        let rotational_info = get_rotational_info(&fs_info.fsname)?;
+        if rotational_info != "0" {
+            warn!("{:?} not on SSD device", data_path);
+        }
+        Ok(())
+    }
+
+    #[cfg(test)]
+    mod test {
+
```
Comment (BusyJay) [inline src/util/config.rs]: Redundant line.

**Comment by BusyJay** [comment]
LGTM, /run-integration-tests

**Comment by AndreMouche** [comment]
@zhangjinpeng1987 PTAL

**Comment by AndreMouche** [comment]
@disksing PTAL

**Comment by disksing** [comment]
LGTM.

**Pull request merged**